### DISCUSSION TITLE
When I call `convert_to_pdf` it causes the resulting pdf font to be ghosted.

### DISCUSSION THREAD

**Discussion by izerui** [pymupdf/PyMuPDF#3356] [Looking for help] [answered]
When I call `convert_to_pdf` it causes the resulting pdf font to be ghosted.

result is: 
![WX20240407-162353@2x](https://github.com/pymupdf/PyMuPDF/assets/1773233/7381e2c0-23df-4ad2-9e6e-ffde9eb60686)

```
    def test_ghost(self):
        bytes = httpx.get(
            'https://tfile.yj2025.com/pdf-processor/source/2024-04-07/mt_04_24024_0_812--1.pdf').content
        doc = fitz.open('pdf', bytes)
        for page in doc:
            page.clean_contents()
        doc = fitz.open('pdf', doc.convert_to_pdf())
        doc.save('xxx.pdf')
```

May I ask, is there a problem, or do I need to convert the time to set what parameters to achieve, it seems that pymupdf in the conversion of the time occurs ocr
thanks!!!

**Comment by JorjMcKie** [comment]
As documented under method `.convert_to_pdf()`, this is not guaranteed to always work. Frequent causes are font errors - which also happens in this case:
```python
In [1]: import fitz
In [2]: doc = fitz.open("test.pdf")
In [3]: pdfdata = doc.convert_to_pdf()
cannot create ToUnicode mapping for NQAVHI+AdobeSongStd-Light
cannot create ToUnicode mapping for KJRDRO+FangSong_GB2312
cannot create ToUnicode mapping for LNBJFM+Symbol_ASME
cannot create ToUnicode mapping for PPUUMW+SimHei
In [4]: new = fitz.open("pdf", pdfdata)
In [5]: new.ez_save("converted.pdf")
```
Messages are from the MuPDF converter.
The result of the method should not be trusted in cases like this - and weird things happening during the conversion are not worth pursuing.

What are you trying to achieve anyway?

**Comment by izerui** [comment]
@JorjMcKie 
I want to copy all the information in a pdf, including the annotations can be copied in the area of another pdf, because `show_pdf_page() ` does not support the copy of the annotations, which may lead to the destination pdf to lose some of the necessary information, I use `convert_to_pdf `, I can get a new pdf, and will be able to convert the annotations into a method that can allow the annotations to work in the ` show_pdf_page`, only the same duplicate lines as ocr, do you know if there are other programs to meet this need? show_pdf_page`, just the same duplicate lines as ocr, I wonder if you have any other program to meet this need?  Thank you very much!

**Reply by JorjMcKie** [reply, marked as answer]
Just as I suspected! I have good news for you:

The base library has a solution for this, which is immediately available in PyMuPDF. There is a function that "bakes" annotations and fields (!!!) into the PDF - which means it converts these items into normal page content.
You have to bake the source PDF before using it in `.show_pdf_page()`. After baking, no annotations and no fields will exist anymore - but otherwise, every page will look exactly equal.

This is how it works:
```python
src = fitz.open("source.pdf")  # this actually is a "fz_document" with an underlying PDF document
src_pdf = fitz.mupdf.pdf_document_from_fz_document(doc)  # access underlying PDF
fitz.mupdf.pdf_bake_document(src_pdf, 1, 1)  # bake annots and fields into the pages

# now the pages of src can be used in `show_pdf_page`.
```
The previous annotations and fields are visible in the target page as normal text (and drawings where applicable), and can thus be extracted like normal.

**Reply by izerui** [reply]
I went out for a meal and came back to such great news, thank you so much, I'm going to try it right away, thanks again!!!!

**Reply by izerui** [reply]
As you say, it works and is perfect.
